### PR TITLE
Pagination

### PR DESCRIPTION
**Pull request by stpCollabr8nLstn**
Add new Pager component as a more responsive pagination component. Pagination remains intact to prevent breakage for current consumers.

#### Changelog

**New**

- Pager/Pager.js
- Pager/Pager-story.js
- Pager/_pager.scss
- Pager/index.js
- Pager/Pager-test.js

### CODE REVIEW

File: src/components/Pagination/Pagination.js [hunk 1]
```diff
@@ -140,41 +243,70 @@ export default class Pagination extends Component {
...
     const classNames = classnames({
-      'bx--pagination': true,
+      'bx--pager': true,
       [className]: className,
     });
-    const tabItems = this.renderTabItems();
+
+    const backwardIcon = {
+      name: 'left',
+      description: 'previous page',
```
Comment (j1mie) [inline src/components/Pagination/Pagination.js]: Would consider putting these in as props for globalisation reasons!

File: src/components/Pagination/_pagination.scss [hunk 2]
```diff
@@ -1,98 +1,127 @@
 @import '../../globals/theme/ics-theme';
-@import 'carbon-components/src/components/pagination/pagination';
```
Comment (j1mie) [inline src/components/Pagination/_pagination.scss]: <3

File: src/components/Pagination/Pagination.js [hunk 3]
```diff
@@ -1,137 +1,240 @@
...
+                </li>
+              );
+            })}
+        </ul>
+      );
+    } else {
+      // build pager layout for middle page grouping
+      this.state.pageQueue = (
+        <ul className="bx--pager__page-list">
+          {this.state.pages.slice(this.state.page - 3, this.state.page + 2).map(page => {
```
Comment (j1mie) [inline src/components/Pagination/Pagination.js]: I'd look at this for the strange paging behaviour

File: src/components/Pagination/Pagination.js [hunk 4]
```diff
@@ -1,137 +1,240 @@
...
     page: PropTypes.number,
     totalItems: PropTypes.number,
+    max: PropTypes.number,
+    mid: PropTypes.number,
+    onClick: PropTypes.func,
+    onKeyUp: PropTypes.func,
   };
 
   static defaultProps = {
-    backwardText: 'Backward',
```
Comment (j1mie) [inline src/components/Pagination/Pagination.js]: Could be worth keeping for globalisation

File: src/components/Pagination/Pagination.js [hunk 5]
```diff
@@ -1,137 +1,240 @@
...
+    }
   };
 
-  decrementPage = () => {
-    const page = this.state.page - 1;
-    this.setState({ page });
-    const tab = this[`tab${page}`];
-    tab.tabAnchor.focus();
-    this.props.onChange({ page });
+  onKeyUp = e => {
```
Comment (j1mie) [inline src/components/Pagination/Pagination.js]: Nice!

File: src/components/Pagination/_pagination.scss [hunk 6]
```diff
@@ -1,98 +1,127 @@
...
     display: inline-block;
-    padding: rem(4px);
-    margin: 0 rem(8px);
+    padding: 0 rem(4px);
+    margin: 0 rem(5px);
     border-radius: 4px;
     border: 1px solid transparent;
+
+    &:focus {
+      outline: none;
```
[the remark this answers is not in the document]
Comment (stpCollabr8nLstn) [reply src/components/Pagination/_pagination.scss]: Can you clarify on this one?

File: src/components/Pager/Pager.js [hunk 7]
```diff
@@ -0,0 +1,299 @@
...
+    forwardText: PropTypes.string,
+  };
+
+  static defaultProps = {
+    onClick: () => {},
+    onKeyUp: () => {},
+    initialPage: 1,
+    max: 5,
+    mid: 3, // number of pages to show on lower/higher ends
+    backwardText: 'previous page',
```
Comment (j1mie) [inline src/components/Pager/Pager.js]: 👍 Good to see these as props!

File: src/components/Pager/Pager.js [hunk 8]
```diff
@@ -0,0 +1,299 @@
...
+    forwardText: 'next page',
+  };
+
+  updatePageQueue = () => {
```
Comment (j1mie) [inline src/components/Pager/Pager.js]: This is so much more concise

File: src/components/Pager/_pager.scss [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+    }
+
+    &:focus {
+      outline: none;
+      box-shadow: none;
+      border: none;
+    }
+  }
+
+  .#{$prefix}--pager__button-icon {
```
Comment (j1mie) [inline src/components/Pager/_pager.scss]: Should probably add a focus style in here too.

File: src/components/Pager/Pager.js [hunk 10]
```diff
@@ -0,0 +1,299 @@
...
+      showCenter: initialPage > mid,
+      showHigher: initialPage >= totalItems - 2,
+    };
+  }
+
+  static propTypes = {
+    className: PropTypes.string,
+    initialPage: PropTypes.number,
+    totalItems: PropTypes.number,
+    max: PropTypes.number,
```
Comment (j1mie) [inline src/components/Pager/Pager.js]: Does this relate solely to truncation? If so, again maybe we could name the prop around truncation here for clarity.

File: src/components/Pager/_mixins.scss [hunk 11]
```diff
@@ -0,0 +1,11 @@
+$hint-transition: all 0.35s cubic-bezier(0.34, -0.61, 1, 0.64);
```
[the remark this answers is not in the document]
Comment (stpCollabr8nLstn) [reply src/components/Pager/_mixins.scss]: good point!

File: src/components/Pager/_pager.scss (hunk 9 above)
Comment (stpCollabr8nLstn) [reply src/components/Pager/_pager.scss]: added it to line 85

File: src/components/Pager/Pager.js (hunk 8 above)
Comment (stpCollabr8nLstn) [reply src/components/Pager/Pager.js]: 🦄

File: src/components/Pager/Pager.js (hunk 10 above)
Comment (stpCollabr8nLstn) [reply src/components/Pager/Pager.js]: 👍

File: src/components/Pager/Pager.js [hunk 12]
```diff
@@ -0,0 +1,299 @@
...
+  }
+}
+
+export default class Pager extends Component {
+  constructor(props) {
+    super(props);
+    const { totalItems, initialPage, mid } = this.props;
+    const lower = [2, 3];
+    const higher = [totalItems - 2, totalItems - 1];
+    const isSmall = totalItems <= this.props.max;
```
[the remark this answers is not in the document]
Comment (stpCollabr8nLstn) [reply src/components/Pager/Pager.js]: 👍

File: src/components/Pager/_pager.scss [hunk 13]
```diff
@@ -0,0 +1,127 @@
...
+    border: none;
+    @include type-scale-item(c);
+  }
+
+  .#{$prefix}--pager__center {
+    display: flex;
+    flex: 1;
+    justify-content: center;
+    flex-direction: column;
+    position: absolute;
```
[the remark this answers is not in the document]
Comment (stpCollabr8nLstn) [reply src/components/Pager/_pager.scss]: will address in future sprint

File: src/components/Pager/_pager.scss [hunk 14]
```diff
@@ -0,0 +1,127 @@
...
+    max-width: rem(186px);
+    @include bidi-left(rem(49px));
+  }
+
+  .#{$prefix}--pager__left {
+    position: absolute;
```
[the remark this answers is not in the document]
Comment (stpCollabr8nLstn) [reply src/components/Pager/_pager.scss]: will address in future sprint

File: src/components/Pager/_pager.scss [hunk 15]
```diff
@@ -0,0 +1,127 @@
...
+    @include bidi-left(0);
+    @include bidi-margin-right(15px);
+  }
+
+  .#{$prefix}--pager__right {
+    position: absolute;
```
[the remark this answers is not in the document]
Comment (stpCollabr8nLstn) [reply src/components/Pager/_pager.scss]: will address in future sprint

**Pull request merged**